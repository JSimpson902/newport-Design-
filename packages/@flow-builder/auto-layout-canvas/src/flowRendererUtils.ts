import ConnectorType from './ConnectorTypeEnum';
import ConnectorLabelType from './ConnectorLabelTypeEnum';
import { SvgInfo, Geometry } from './svgUtils';
import {
    FlowModel,
    ElementsMetadata,
    ElementMetadata,
    NodeRef,
    NodeModel,
    BranchHeadNodeModel,
    ParentNodeModel,
    Guid
} from './model';
import MenuType from './MenuType';

export interface LayoutInfo {
    x: number;
    y: number;
    w: number;
    h: number;
    joinOffsetY: number;
    offsetX: number;
    addOffset: number;
    labelOffset: number;
}

export interface NodeLayout {
    prevLayout?: LayoutInfo | undefined;
    layout: LayoutInfo;
}

export interface NodeLayoutMap {
    [key: string]: NodeLayout;
}

export type NodeDimensionMap = Map<Guid, Dimension>;

export interface FlowRenderContext {
    flowModel: FlowModel;
    nodeLayoutMap: NodeLayoutMap;
    progress: number;
    interactionState: FlowInteractionState;
    elementsMetadata: ElementsMetadata;
    layoutConfig: LayoutConfig;
    isFault: boolean;
    isDeletingBranch: boolean;
    dynamicNodeDimensionMap: NodeDimensionMap;
}

export interface InteractionMenuInfo {
    key: Guid;
    type: MenuType;
    geometry?: Geometry;
    needToPosition: boolean;
}

export interface FlowInteractionState {
    menuInfo: InteractionMenuInfo | null;
    deletionPathInfo: {
        childIndexToKeep: number;
        elementGuidToDelete: Guid;
    } | null;
}

export interface FlowRenderInfo {
    geometry: Geometry;
    nodes: NodeRenderInfo[];
    isTerminal: boolean;
    layoutConfig: object;
    preConnector?: ConnectorRenderInfo;
}

export interface NodeRenderInfo {
    geometry: Geometry;
    label: string;
    metadata: ElementMetadata;
    config: { isSelected: boolean; isHighlighted: boolean; isSelectable: boolean; hasError: boolean };
    nextConnector?: ConnectorRenderInfo;
    flows: FlowRenderInfo[];
    faultFlow?: FlowRenderInfo;
    isNew?: boolean;
    logicConnectors: ConnectorRenderInfo[];
    conditionOptions?: Option[];
    defaultConnectorLabel?: string;
    guid: string;
    menuOpened: boolean;
    isTerminal: boolean;
    toBeDeleted: boolean;
    node?: NodeModel;
    dynamicNodeComponent?: string;
    dynamicNodeComponentSelector?: Function;
}

export interface Dimension {
    w: number;
    h: number;
}

export enum ConnectorVariant {
    DEFAULT = 'default',
    EDGE = 'edge',
    EDGE_BOTTOM = 'edgeBottom',
    CENTER = 'center',
    LOOP = 'loop',
    FAULT = 'fault',
    POST_MERGE = 'postMerge',
    POST_MERGE_TAIL = 'postMergeTail',
    BRANCH_TAIL = 'branchTail',
    BRANCH_HEAD = 'branchHead',
    BRANCH_HEAD_EMPTY = 'branchHeadEmpty',
    DEFAULT_LABEL = 'defaultLabel'
}
export enum VerticalAlign {
    TOP,
    BOTTOM
}

export interface ConnectorTypeLayoutConfig {
    h: number;
    addOffset: number;
    labelOffset?: number;
    variants?: {
        [key in ConnectorVariant]?: any;
    };
    svgMarginTop?: number;
    svgMarginBottom?: number;
}

export interface LayoutConfig {
    grid: {
        w: number;
        h: number;
    };
    menu: {
        marginBottom: number;
    };
    connector: {
        icon: {
            w: number;
            h: number;
        };
        strokeWidth: number;
        curveRadius: number;
        menu: Dimension;
        types: {
            [key in ConnectorType]?: ConnectorTypeLayoutConfig;
        };
    };
    node: {
        icon: {
            w: number;
            h: number;
        };
        w: number;
        offsetX: number;
    };
    branch: {
        defaultWidth: number;
        emptyWidth: number;
    };
}

export interface Option {
    label: string;
    value: Guid;
}

export interface MenuInfo {
    guid: NodeRef;
    menuOpened: boolean;
}

export interface ConnectorAddInfo {
    offsetY: number;
    menuOpened: boolean;
}

export interface SelectInfo {
    value: NodeRef;
    options: Option[];
}

export interface ConnectorConnectionInfo {
    prev?: NodeRef;
    next?: NodeRef;
    parent?: NodeRef;
    childIndex?: number;
}

export interface ConnectorRenderInfo {
    type: ConnectorType;
    labelType: ConnectorLabelType;
    geometry: Geometry;
    svgInfo: SvgInfo;
    addInfo?: ConnectorAddInfo;
    connectionInfo: ConnectorConnectionInfo;
    isFault: boolean;
    labelOffsetY?: number;
    connectorBadgeLabel?: string;
    isNew?: boolean;
    toBeDeleted: boolean;
    isHighlighted: boolean;
}

function getBranchLayoutKey(parentGuid: string, childIndex: number) {
    return `${parentGuid}:${childIndex}`;
}

function getLayoutByKey(key: string, progress: number, nodeLayoutMap: NodeLayoutMap): LayoutInfo {
    const { prevLayout, layout } = nodeLayoutMap[key];

    return !prevLayout || progress === 1 ? layout : tween(prevLayout, layout, progress);
}

export function tweenValue(prevValue: number, value: number, progress: number): number {
    const delta = (value - prevValue) * progress;
    return prevValue + delta;
}

/**
 * Tweens LayoutInfo values
 *
 * @param prevLayout - The previous LayoutInfo
 * @param layout - The current LayoutInfo
 * @param progress - The rendering progress
 * @returns A LayoutInfo that with the tweened values
 */
function tween(prevLayout: LayoutInfo, layout: LayoutInfo, progress: number): LayoutInfo {
    const { x, y, w, h, offsetX, joinOffsetY, addOffset, labelOffset } = layout;

    return {
        x: tweenValue(prevLayout.x, x, progress),
        y: tweenValue(prevLayout.y, y, progress),
        w: tweenValue(prevLayout.w, w, progress),
        h: tweenValue(prevLayout.h, h, progress),
        offsetX: tweenValue(prevLayout.offsetX, offsetX, progress),
        joinOffsetY: tweenValue(prevLayout.joinOffsetY, joinOffsetY, progress),
        addOffset: prevLayout.addOffset !== 0 ? tweenValue(prevLayout.addOffset, addOffset, progress) : addOffset,
        labelOffset:
            prevLayout.labelOffset !== 0 ? tweenValue(prevLayout.labelOffset, labelOffset, progress) : labelOffset
    };
}

function createDefaultLayout(): LayoutInfo {
    return {
        x: 0,
        y: 0,
        w: 0,
        h: 0,
        joinOffsetY: 0,
        offsetX: 0,
        addOffset: 0,
        labelOffset: 0
    };
}

function getLayout(nodeGuid: string, progress: number, nodeLayoutMap: NodeLayoutMap): LayoutInfo {
    if (!nodeLayoutMap[nodeGuid]) {
        return createDefaultLayout();
    }
    return getLayoutByKey(nodeGuid, progress, nodeLayoutMap);
}

/**
 * Get a LayoutInfo for a branch
 *
 * @param parentGuid - A branching element guid
 * @param childIndex  - A branch index
 * @param progress - Rendering progress
 * @param nodeLayoutMap - The node layout map
 * @returns A LayoutInfo for the branch
 */
function getBranchLayout(
    parentGuid: Guid,
    childIndex: number,
    progress: number,
    nodeLayoutMap: NodeLayoutMap
): LayoutInfo {
    const key = getBranchLayoutKey(parentGuid, childIndex);
    if (!nodeLayoutMap[key]) {
        return createDefaultLayout();
    }
    return getLayoutByKey(key, progress, nodeLayoutMap);
}

/**
 * Get the config for a connector type and variants
 *
 * @param layoutConfig - The layout config
 * @param connectorType - The connector type
 * @param connectorVariants - A list of variants
 * @return A connector layout config for the connector
 */
function getConnectorConfig(
    layoutConfig: LayoutConfig,
    connectorType: ConnectorType,
    ...connectorVariants: ConnectorVariant[]
): ConnectorTypeLayoutConfig {
    let connectorConfig = layoutConfig.connector.types[connectorType]!;

    // go through variants and override values
    connectorVariants.forEach((connectorVariant: ConnectorVariant) => {
        const variantConfig = connectorConfig.variants![connectorVariant];
        if (variantConfig != null) {
            connectorConfig = { ...connectorConfig, ...variantConfig } as ConnectorTypeLayoutConfig;
        }
    });

    return connectorConfig;
}

/**
 * For a branching element, get the number of outcomes that merge back
 *
 * @param flowModel - The flow model
 * @param branchingElement - THe branching element
 * @return the number of outcomes that merge back
 */
const getMergeOutcomeCount = (flowModel: FlowModel, branchingElement: ParentNodeModel) =>
    branchingElement.children.reduce((count, child) => {
        return child == null || !(flowModel[child] as BranchHeadNodeModel).isTerminal ? count + 1 : count;
    }, 0);

export {
    getBranchLayoutKey,
    getBranchLayout,
    tween,
    createDefaultLayout,
    getLayout,
    getConnectorConfig,
    getMergeOutcomeCount
};
