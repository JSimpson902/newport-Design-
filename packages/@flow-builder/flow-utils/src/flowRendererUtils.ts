import { LayoutInfo, NodeLayoutMap, getBranchLayoutKey } from './layout';
import ConnectorType from './ConnectorTypeEnum';
import { SvgInfo, Geometry } from './svgUtils';
import { FlowModel, ElementsMetadata, ElementMetadata, NodeRef, Guid } from './model';
import MenuType from './MenuType';
import ElementType from './ElementType';

export enum ConditionType {
    DEFAULT,
    FAULT,
    STANDARD
}

export interface FlowRenderContext {
    flowModel: FlowModel;
    nodeLayoutMap: NodeLayoutMap;
    progress: number;
    interactionState: FlowInteractionState;
    elementsMetadata: ElementsMetadata;
    layoutConfig: LayoutConfig;
    isFault: boolean;
}

export interface FlowInteractionState {
    menuInfo: {
        key: Guid;
        type: MenuType;
    } | null;
}

export interface FlowRenderInfo {
    geometry: Geometry;
    nodes: NodeRenderInfo[];
    isTerminal: boolean;
    layoutConfig: object;
    preConnector?: ConnectorRenderInfo;
}

export interface NodeRenderInfo {
    geometry: Geometry;
    label: string;
    metadata: ElementMetadata;
    config: { isSelected: boolean; isHighlighted: boolean; canSelect: boolean };
    nextConnector?: ConnectorRenderInfo;
    flows: FlowRenderInfo[];
    faultFlow?: FlowRenderInfo;
    isNew?: boolean;
    logicConnectors: ConnectorRenderInfo[];
    conditionOptions?: Option[];
    defaultConnectorLabel?: string;
    guid: string;
    menuOpened: boolean;
    isTerminal: boolean;
}

export interface Dimension {
    w: number;
    h: number;
}

export enum ConnectorVariant {
    DEFAULT = 'default',
    EDGE = 'edge',
    CENTER = 'center'
}

interface ConnectorTypeLayoutConfig {
    h: number;
    addOffset: number;
    labelOffset?: number;
    variants?: {
        [key in ConnectorVariant]: any;
    };
    svgMarginTop?: number;
    svgMarginBottom?: number;
}

export interface LayoutConfig {
    grid: {
        w: number;
        h: number;
    };
    menu: {
        marginBottom: number;
    };
    connector: {
        icon: {
            w: number;
            h: number;
        };
        strokeWidth: number;
        curveRadius: number;
        menu: Dimension;
        types: {
            [key in ConnectorType]?: ConnectorTypeLayoutConfig;
        };
    };
    node: {
        icon: {
            w: number;
            h: number;
        };
        w: number;
        offsetX: number;
        menu: {
            [key in ElementType]?: Dimension | undefined;
        };
    };
    branch: {
        defaultWidth: number;
        emptyWidth: number;
    };
}

export interface Option {
    label: string;
    value: Guid;
}

export interface MenuInfo {
    guid: NodeRef;
    menuOpened: boolean;
}

export interface ConnectorAddInfo {
    offsetY: number;
    menuOpened: boolean;
}

export interface SelectInfo {
    value: NodeRef;
    options: Option[];
}

export interface ConnectorConnectionInfo {
    prev?: NodeRef;
    next?: NodeRef;
    parent?: NodeRef;
    childIndex?: number;
}

export interface ConnectorRenderInfo {
    type: ConnectorType;
    geometry: Geometry;
    svgInfo: SvgInfo;
    addInfo?: ConnectorAddInfo;
    connectionInfo: ConnectorConnectionInfo;

    isFault: boolean;

    labelOffsetY?: number;
    conditionOptions?: Option[];
    conditionValue?: Guid;
    conditionType?: ConditionType;
    isNew?: boolean;
}

function getLayoutByKey(key: string, progress: number, nodeLayoutMap: NodeLayoutMap): LayoutInfo {
    const { prevLayout, layout } = nodeLayoutMap[key];

    return !prevLayout || progress === 1 ? layout : tween(prevLayout, layout, progress);
}

function tweenValue(prevValue: number, value: number, progress: number): number {
    const delta = (value - prevValue) * progress;
    return prevValue + delta;
}

/**
 * Tweens LayoutInfo values
 *
 * @param prevLayout - The previous LayoutInfo
 * @param layout - The current LayoutInfo
 * @param progress - The rendering progress
 * @returns A LayoutInfo that with the tweened values
 */
function tween(prevLayout: LayoutInfo, layout: LayoutInfo, progress: number): LayoutInfo {
    return {
        x: tweenValue(prevLayout.x, layout.x, progress),
        y: tweenValue(prevLayout.y, layout.y, progress),
        w: tweenValue(prevLayout.w, layout.w, progress),
        h: tweenValue(prevLayout.h, layout.h, progress),
        offsetX: tweenValue(prevLayout.offsetX, layout.offsetX, progress),
        joinOffsetY: tweenValue(prevLayout.joinOffsetY, layout.joinOffsetY, progress)
    };
}

function getLayout(nodeGuid: string, progress: number, nodeLayoutMap: NodeLayoutMap): LayoutInfo {
    return getLayoutByKey(nodeGuid, progress, nodeLayoutMap);
}

/**
 * Get a LayoutInfo for a branch
 *
 * @param parentGuid - A branching element guid
 * @param childIndex  - A branch index
 * @param progress - Rendering progress
 * @param nodeLayoutMap - The node layout map
 * @returns A LayoutInfo for the branch
 */
function getBranchLayout(
    parentGuid: Guid,
    childIndex: number,
    progress: number,
    nodeLayoutMap: NodeLayoutMap
): LayoutInfo {
    const key = getBranchLayoutKey(parentGuid, childIndex);
    return getLayoutByKey(key, progress, nodeLayoutMap);
}

export { getBranchLayout, tween, getLayout };
