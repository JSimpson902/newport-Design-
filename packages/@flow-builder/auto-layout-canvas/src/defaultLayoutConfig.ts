import { LayoutConfig } from './flowRendererUtils';

/**
 * Returns the default config for the layout algorithm
 *
 * @returns The default LayoutConfig for the layout
 */
export function getDefaultLayoutConfig(): LayoutConfig {
    const gridHeight = 24;
    const gridWidth = 132;

    return {
        grid: {
            w: gridWidth,
            h: gridHeight
        },
        menu: {
            marginBottom: gridHeight
        },
        connector: {
            icon: {
                w: 20,
                h: 20
            },
            strokeWidth: 6,
            curveRadius: 16,
            menu: {
                w: 100,
                h: 328
            },
            types: {
                straight: {
                    addOffset: gridHeight * 2.5,
                    h: gridHeight * 5,
                    labelOffset: gridHeight * 2,
                    variants: {
                        fault: {
                            labelOffset: gridHeight,
                            svgMarginTop: gridHeight
                        },
                        postMerge: {
                            addOffset: gridHeight * 1.5,
                            h: 4 * gridHeight,
                            svgMarginTop: gridHeight,
                            variants: {
                                edge: { svgMarginBottom: gridHeight },
                                center: { svgMarginBottom: -gridHeight }
                            }
                        },
                        postMergeTail: {
                            addOffset: gridHeight * 1.5,
                            h: 3 * gridHeight,
                            svgMarginTop: gridHeight,
                            variants: {
                                edge: { svgMarginBottom: gridHeight },
                                center: { svgMarginBottom: -gridHeight },
                                loop: {
                                    svgMarginBottom: 2 * gridHeight
                                }
                            }
                        },
                        branchTail: {
                            h: 4 * gridHeight,
                            variants: {
                                edge: {
                                    svgMarginBottom: gridHeight
                                },
                                center: { svgMarginBottom: -gridHeight },
                                loop: {
                                    svgMarginBottom: 2 * gridHeight
                                }
                            }
                        },
                        branchHead: {
                            addOffset: gridHeight * 3.5,
                            h: 6 * gridHeight,
                            variants: {
                                edge: {
                                    svgMarginTop: gridHeight
                                },
                                center: { svgMarginBottom: -gridHeight },
                                fault: {
                                    labelOffset: gridHeight,
                                    svgMarginTop: gridHeight
                                }
                            }
                        },
                        branchHeadEmpty: {
                            addOffset: gridHeight * 3.5,
                            h: 5 * gridHeight,
                            variants: {
                                edge: {
                                    svgMarginTop: gridHeight,
                                    svgMarginBottom: gridHeight
                                },
                                loop: {
                                    svgMarginBottom: 2 * gridHeight
                                },
                                center: { svgMarginBottom: -gridHeight },
                                fault: {
                                    labelOffset: gridHeight,
                                    svgMarginTop: gridHeight
                                }
                            }
                        }
                    }
                }
            }
        },
        node: {
            icon: {
                w: 48,
                h: 48
            },
            offsetX: (gridWidth * 2) / 3,
            w: gridWidth * 2,
            menu: {
                default: {
                    w: 100,
                    h: 225
                },
                branch: {
                    w: 100,
                    h: 225
                },
                loop: {
                    w: 100,
                    h: 225
                },
                start: {
                    w: 100,
                    h: 200
                }
            }
        },
        branch: {
            defaultWidth: gridWidth * 2,
            emptyWidth: (gridWidth * 4) / 3
        }
    };
}
