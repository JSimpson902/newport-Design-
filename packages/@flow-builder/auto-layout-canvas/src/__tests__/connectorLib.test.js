import {
    createBranchConnector,
    createMergeConnector,
    createConnectorToNextNode,
    createLoopBackConnector,
    createLoopAfterLastConnector
} from '../connectorLib';
import ConnectorType from '../ConnectorTypeEnum';
import { ConnectorVariant } from '../flowRendererUtils';
import { getDefaultLayoutConfig } from '../defaultLayoutConfig';
import ConnectorLabelType from '../ConnectorLabelTypeEnum';

jest.mock('../svgUtils.ts', () => {
    return {
        createSvgInfo: (svgPathParams, startOffset) => ({ svgPathParams, startOffset }),
        createOffsetLocation: (location, offset) => ({ location, offset }),
        createSvgPath: (pathParams, startOffset) => ({ pathParams, startOffset })
    };
});

describe('connectorLib', () => {
    it('createLoopBackConnector', () => {
        const connectorRenderInfo = createLoopBackConnector(
            'guid',
            { x: 0, y: 0, w: 88, h: 168 },
            getDefaultLayoutConfig(),
            false,
            false
        );

        expect(connectorRenderInfo).toMatchSnapshot();
    });

    it('createLoopAfterLastConnector', () => {
        const connectorRenderInfo = createLoopAfterLastConnector(
            'guid',
            { x: -88, y: 0, w: 88, h: 168 },
            getDefaultLayoutConfig(),
            false,
            false
        );

        expect(connectorRenderInfo).toMatchSnapshot();
    });

    it('createBranchConnector', () => {
        const connectorRenderInfo = createBranchConnector(
            'guid',
            0,
            { x: 0, y: 0, w: 100, h: 28 },
            ConnectorType.BRANCH_LEFT,
            getDefaultLayoutConfig(),
            false
        );

        expect(connectorRenderInfo).toMatchSnapshot();
    });

    it('createMergeConnector', () => {
        const connectorRenderInfo = createMergeConnector(
            'guid',
            0,
            { x: 0, y: 0, w: 100, h: 28 },
            ConnectorType.MERGE_LEFT,
            getDefaultLayoutConfig(),
            false
        );

        expect(connectorRenderInfo).toMatchSnapshot();
    });

    it('createConnectorToNextNode', () => {
        const connectorRenderInfo = createConnectorToNextNode(
            { next: 'nextGuid', prev: 'prevGuid' },
            ConnectorType.STRAIGHT,
            ConnectorLabelType.NONE,
            0,
            144,
            false,
            getDefaultLayoutConfig(),
            false,
            ConnectorVariant.DEFAULT
        );

        expect(connectorRenderInfo).toMatchSnapshot();
    });
});
