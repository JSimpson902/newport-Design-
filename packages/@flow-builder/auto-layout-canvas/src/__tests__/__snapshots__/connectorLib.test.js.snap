// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectorLib createBranchConnector 1`] = `
Object {
  "connectionInfo": Object {
    "childIndex": 0,
    "parent": "guid",
  },
  "geometry": Object {
    "h": 28,
    "w": 100,
    "x": 0,
    "y": 0,
  },
  "isFault": false,
  "labelType": 0,
  "svgInfo": Object {
    "startOffset": Array [
      3,
      3,
    ],
    "svgPathParams": Object {
      "offsets": Array [
        Array [
          -84,
          0,
        ],
        Array [
          -16,
          16,
        ],
        Array [
          0,
          12,
        ],
      ],
      "start": Object {
        "x": 100,
        "y": 0,
      },
    },
  },
  "toBeDeleted": undefined,
  "type": "branchLeft",
}
`;

exports[`connectorLib createConnectorToNextNode 1`] = `
Object {
  "addInfo": undefined,
  "connectionInfo": Object {
    "next": "nextGuid",
    "prev": "prevGuid",
  },
  "geometry": Object {
    "h": 144,
    "w": 6,
    "x": 0,
    "y": 0,
  },
  "isFault": false,
  "labelOffsetY": 48,
  "labelType": 0,
  "svgInfo": Object {
    "geometry": Object {
      "h": 144,
      "w": 6,
      "x": -3,
      "y": 0,
    },
    "path": Object {
      "pathParams": Object {
        "offsets": Array [
          Array [
            0,
            144,
          ],
        ],
        "start": Object {
          "location": Object {
            "x": 0,
            "y": 0,
          },
          "offset": Array [
            3,
            0,
          ],
        },
      },
      "startOffset": undefined,
    },
  },
  "toBeDeleted": undefined,
  "type": "straight",
}
`;

exports[`connectorLib createLoopAfterLastConnector 1`] = `
Object {
  "connectionInfo": Object {
    "parent": "guid",
  },
  "geometry": Object {
    "h": 168,
    "w": 88,
    "x": -88,
    "y": 0,
  },
  "isFault": false,
  "labelOffsetY": 48,
  "labelType": 3,
  "svgInfo": Object {
    "startOffset": Array [
      3,
      3,
    ],
    "svgPathParams": Object {
      "offsets": Array [
        Array [
          -72,
          0,
        ],
        Array [
          -16,
          16,
        ],
        Array [
          0,
          136,
        ],
        Array [
          16,
          16,
        ],
        Array [
          56,
          0,
        ],
        Array [
          16,
          16,
        ],
        Array [
          0,
          8,
        ],
      ],
      "start": Object {
        "x": 88,
        "y": 0,
      },
    },
  },
  "toBeDeleted": false,
  "type": "loopAfterLast",
}
`;

exports[`connectorLib createLoopBackConnector 1`] = `
Object {
  "connectionInfo": Object {
    "parent": "guid",
  },
  "geometry": Object {
    "h": 168,
    "w": 88,
    "x": 0,
    "y": 0,
  },
  "isFault": false,
  "labelType": 0,
  "svgInfo": Object {
    "startOffset": Array [
      3,
      3,
    ],
    "svgPathParams": Object {
      "offsets": Array [
        Array [
          0,
          8,
        ],
        Array [
          16,
          16,
        ],
        Array [
          56,
          0,
        ],
        Array [
          16,
          -16,
        ],
        Array [
          0,
          -112,
        ],
        Array [
          -16,
          -16,
        ],
        Array [
          -72,
          0,
        ],
      ],
      "start": Object {
        "x": 0,
        "y": 120,
      },
    },
  },
  "toBeDeleted": false,
  "type": "loopBack",
}
`;

exports[`connectorLib createMergeConnector 1`] = `
Object {
  "connectionInfo": Object {
    "childIndex": 0,
    "parent": "guid",
  },
  "geometry": Object {
    "h": 28,
    "w": 100,
    "x": 0,
    "y": 0,
  },
  "isFault": false,
  "labelType": 0,
  "svgInfo": Object {
    "startOffset": Array [
      3,
      0,
    ],
    "svgPathParams": Object {
      "offsets": Array [
        Array [
          0,
          -2,
        ],
        Array [
          16,
          16,
        ],
        Array [
          68,
          0,
        ],
        Array [
          16,
          16,
        ],
        Array [
          0,
          -2,
        ],
      ],
      "start": Object {
        "x": 0,
        "y": 0,
      },
    },
  },
  "toBeDeleted": undefined,
  "type": "mergeLeft",
}
`;
